[DRAFT] Support changelog scan for table with delete files

wypoon opened the pull request:
Currently changelog scan is only supported for a table with no delete files. We implement support for the case when delete files are present in the snapshots to be scanned.

pvary commented on core/src/main/java/org/apache/iceberg/BaseIncrementalChangelogScan.java:
```diff
@@ -133,51 +131,155 @@ private static Map<Long, Integer> computeSnapshotOrdinals(Deque<Snapshot> snapsh
+                    } else {
+                      // the data file is added before the snapshot we're processing
+                      if (addedDeleteFiles.length == 0) {
+                        return new DummyChangelogScanTask(changeOrdinal, snapshotId);
```
nit: why don't you have a static object here?

wypoon replied:
Thanks for the suggestion.

wypoon replied:
Adopted your suggestion to use a static instance for DummyChangelogScanTask.

pvary commented on core/src/main/java/org/apache/iceberg/BaseIncrementalChangelogScan.java:
```diff
@@ -133,51 +131,155 @@ private static Map<Long, Integer> computeSnapshotOrdinals(Deque<Snapshot> snapsh
+                    if (addedDeleteFiles.length == 0) {
+                      return new DummyChangelogScanTask(changeOrdinal, snapshotId);
+                    } else {
+                      return new BaseDeletedRowsScanTask(
```
Would this cause a full table scan for every snapshot in the snapshot range we are checking when there is an equality delete file present?

wypoon replied:
Note that a BaseDeletedRowsScanTask is only created if there are delete files *added in that snapshot* that may apply to the data file.
Supposing that there is an equality delete file added in that snapshot, whether a full table scan happens for that snapshot depends on the equality deletes and how well `DeleteFileIndex.canContainEqDeletesForFile(DataFile, EqualityDeleteFile)` is able to filter them. If the equality deletes are for a partition, or even if global but can be excluded using ranges, then a full table scan may not happen. But yes, depending on the equality deletes it could happen.
Unfortunately, I don't see an alternative.

manuzhang commented:
@wypoon have you checked out https://github.com/apache/iceberg/pull/9888?

wypoon commented:
> @wypoon have you checked out #9888?

@manuzhang I was not aware of it. Let me take a look. Unfortunately a lot of PRs don't get reviewed. All I knew was that changelog scan for the case where delete files are present was not implemented in main up to now.

wypoon commented:
@manuzhang thank you for pointing me to your PR, https://github.com/apache/iceberg/pull/9888. I have tried testing it and left comments on your PR.

wypoon commented:
@aokolnychyi @flyrain @stevenzwu @szehon-ho can you please review this?

pvary commented:
@wypoon, @manuzhang: I'm interested in providing Flink streaming CDC reads. That would require a working changelog scan planning in Iceberg core. So I would be happy to help with my reviews here. I'm not an expert of this part of the code yet, so if the PR becomes more complicated, we might have to ask help from folks more experienced here, but I can do a first round of the reviews.

I see that there are multiple PRs for this topic (#9888, #10954). I would suggest to focus the discussion on one specific PR.

There are some tricky scenarios with Flink CDC, we need to handle:

-----------------------

Scenario1:
- 1st snapshot: Writing a new record to the table primary key 1 (PK1), and value 1 (V1)
    - Creates data file 1  (DF1) - with PK1-V1
- 2nd snapshot: The record is updated with PK1 with value V2
    - Creates an equality delete 1 (ED1) - with PK1
    - Creates a new data file 2 (DF2) - with PK1-V2
- 3rd snapshot:  The record is updated with PK1 with value V3
    - Creates an equality delete 2 (ED2) - with PK1
    - Creates a new data file 3 (DF3) - with PK1-V3

When the changelog scan reads the 3rd snapshot, it should consider:
- Delete reads:
    - DF1 - omit records deleted by ED1, emit records deleted by ED2
    - DF2 - emit records deleted by ED2
- Data read:
    - DF3 - emit all records

Notice that for DF1 we should not emit records which are deleted by previous deletes. So I am concerned about your `// not used` comment [here](https://github.com/apache/iceberg/pull/10935/files#diff-24a8b4b8a16e20a7727ca9b04754fa00cbb40228aa79b3e63baf958aef7bec04R266) 😄

-----------------------

Scenario2:
- 1st snapshot: Writing a new record to the table primary key 1 (PK1), and value 1 (V1)
    - Creates data file 1  (DF1) - with PK1-V1
- 2nd snapshot: The record is updated with PK1 with value V2, but later updated again with PK1 with value V3
    - Creates an equality delete 1 (ED1) - with PK1
    - Creates a new data file 2 (DF2) - with PK1-V2
    - Creates a positional delete 1 (PS1) - for the DF2-PK1
    - Here we have 2 possibilities:
         - Adds a new line to DF2 with PK1-V3 - if the data file target file size is not reached yet
         - Creates a new data file 3 (DF3)  - with PK1-V3 - if the data file is already rolled over

When changelog scan the 2nd snapshot, it should consider:
- Delete reads:
    - DF1 - emit records deleted by ED1 - the emitted record is: D(PK1-V1)
    - DF2 - emit records deleted by PD1 - the emitted record is: D(PK1-V2)
- Data read:
    - DF2 - omit records deleted by PS1 - the emitted record is: I(PK1-V2)
    - DF3 - emit all records - the emitted record is: I(PK1-V3)

Notice that the order of the records is important, and not trivial to order if the files are read in the distributed way.

CC: @dramaticlly

wypoon commented:
@pvary thank you for your interest and for the Flink scenarios, which is very helpful as I am unfamiliar with Flink.

Regarding https://github.com/apache/iceberg/pull/9888, please read my comments there. I put up https://github.com/apache/iceberg/pull/10954 only as a reference for @manuzhang so he can see the tests I added which fail with the implementation of `BaseIncrementalChangelogScan` in https://github.com/apache/iceberg/pull/9888. https://github.com/apache/iceberg/pull/10954 is not really for consideration or review.

Next week I'll look into the scenarios you listed and see what gaps there are in my implementation and add tests as necessary. Regarding `DeletedRowsScanTask`, when @aokolnychyi added the API, he figured that it was necessary to know what deletes existed before the snapshot and what deletes are added in the snapshot. I'll have to think through the equality delete case, but for the positional delete case, I believe that it is not necessary to know the existing deletes (which is the reason for my `// not used` comment). For the positional delete case, I do not believe deleting the same position in the same data file again can happen, so added deletes should always be new positions. Thus, we only need to scan the data file and emit the rows that are deleted by the added deletes (which I do by using the _pos metadata column and the `PositionDeleteIndex` a `DeleteFilter` constructs for the data file). The _pos metadata column is automatically added to the schema if there are any position delete files to be applied to the data file, and for a `DeletedRowsScanTask`, the `DeleteFilter` is constructed with the added delete files (so those are the position delete files to be applied).
I admit that I hadn't thought through equality deletes carefully. Just to clarify, in your scenario1, for example, does ED1 contain `PK=PK1`? In other words, `PK1` is the value of the primary key in the table, right? and by `V1` you simply mean the values of the other columns? And then ED2 again contains `PK=PK1`?  so that you can update the other columns of the same row with `V2`, right?

pvary commented:
> For the positional delete case, I do not believe deleting the same position in the same data file again can happen

I agree, that this should be the case. My only concern is, that until now, it didn't cause any issues. We might not thought about it and didn't clarify this in the specification - so if we have a hard requirement here, we might want to clarify it in the spec too.

> Just to clarify, in your scenario1, for example, does ED1 contain PK=PK1? In other words, PK1 is the value of the primary key in the table, right? and by V1 you simply mean the values of the other columns? And then ED2 again contains PK=PK1? so that you can update the other columns of the same row with V2, right?

Yes, undestand correctly. That's what I was trying to describe

wypoon commented:
@pvary I have added `testFlinkScenario1` and `testFlinkScenario2` to `TestChangelogReader`. Please check the expected results. (I will rename the tests later with more descriptive names.)
For scenario 1, I agree with you on what the changelog should emit. If DF1 is in snapshot 3, then we should emit the row with PK1 being deleted by ED2. (The row is deleted by ED1 too, but we should only emit the row once, not twice, for snapshot 3).
For scenario 2, I think that when a row in a data file is deleted by a positional delete in the same commit, that row should neither be shown as inserted nor as deleted. This is where I think we disagree. (IIUC, you expect to see it as deleted but not as inserted. To me, that would be inconsistent.) This part of scenario 2 is actually already tested by `testAddingAndDeletingInSameCommit`.
If you agree with my analysis, then my implementation does handle at least your two scenarios correctly.

pvary commented on spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestChangelogReader.java:
```diff
@@ -191,39 +214,361 @@ public void testMixDeleteAndInsert() throws IOException {
+    addExpectedRows(expectedRows, ChangelogOperation.DELETE, snapshotId2, 1, deleted1);
+    addExpectedRows(expectedRows, ChangelogOperation.INSERT, snapshotId2, 1, rl2);
+    // in snapshot 3, eqDeletes2 applies to both df1 and df2, so there are two deletes emitted
+    addExpectedRows(expectedRows, ChangelogOperation.DELETE, snapshotId3, 2, deleted1);
```
I think this one is not correct. This record is not deleted in snapshot3

wypoon replied:
I have fixed the expected results to remove this row.

pvary commented on spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestChangelogReader.java:
```diff
@@ -191,39 +214,361 @@ public void testMixDeleteAndInsert() throws IOException {
+
+    List<InternalRow> rows = getChangelogRows(table2);
+
+    // for snapshot 2, we should record only one delete for id=2 (from the equality delete)
+    // and one insert (the final value)
```
This is one way to present this data. And I would be totally fine with this result. For Flink this is as good as it can get.

Other writers might write delete records into the positional delete records (like `df1.path(), 0L` - to remove "1,a") along with the ones which remove the the new records. If those are handled correctly, then we are golden here.

pvary commented:
> @pvary I have added `testFlinkScenario1` and `testFlinkScenario2` to `TestChangelogReader` based on your two scenarios. Please check the expected results. (I will rename the tests later with more descriptive names.)

These are very great tests. Thanks for implementing them!

> For scenario 1, I agree with you on what the changelog should emit. If DF1 is in snapshot 3, then we should emit the row with PK1 being deleted by ED2. (The row is deleted by ED1 too, but we should only emit the row once, not twice, for snapshot 3).

Do I understand correctly that this is not yet the situation with the current code?

> For scenario 2, I think that when a row in a data file is deleted by a positional delete in the same commit, that row should neither be shown as inserted nor as deleted. This is where I think we disagree. (IIUC, you expect to see it as deleted but not as inserted. To me, that would be inconsistent.) This part of scenario 2 is actually already tested by `testAddingAndDeletingInSameCommit`.

I think we agree here. I'm perfectly fine if we can make sure that the added and immediately removed records are not emitted during the incremental scan read.

> If you agree with my analysis, then my implementation does handle at least your two scenarios correctly.

I think the output for snapshot3 in scenario1 is not correct. I have left a comment there.

wypoon commented:
@pvary thanks for reviewing the tests!
Just a quick response for now. For scenario 1, the behavior is as I thought you described, yet in your comment on the test code, you indicated that it is incorrect, so I am confused. To be honest, I thought about it some more yesterday and I actually think it is incorrect too. So I sent an email to the Iceberg dev list (please see that email), asking for clarification.

pvary commented:
> @pvary thanks for reviewing the tests! Just a quick response for now. For scenario 1, the behavior is as I thought you described, yet in your comment on the test code, you indicated that it is incorrect, so I am confused. To be honest, I thought about it some more yesterday and I actually think it is incorrect too. So I sent an email to the Iceberg dev list (please see that email), asking for clarification. The current behavior is case (b) in the email, which is what I thought you expected. I now think the behavior should be case (a). Please add your thoughts to the thread in the dev list too.

Sorry for the confusion @wypoon!
I have tried to describe the (a) case in my description with this sentence:
> Notice that for DF1 we should not emit records which are deleted by previous deletes.

Anyways, added my comments to the thread as well

wypoon commented:
@pvary thanks for posting on the dev list thread!
In your description of scenario 1, you wrote:

> 

> When the changelog scan reads the 3rd snapshot, it should consider:
> 
> Delete reads:
> DF1 - omit records deleted by ED1, emit records deleted by ED2
> DF2 - emit records deleted by ED2
> Data read:
> DF3 - emit all records

This corresponds to case (b) in the thread. But now I understand that you must have made a mistake and did not really mean the
> DF1 - omit records deleted by ED1, emit records deleted by ED2

and think that (a) is the correct behavior.
The current behavior is (b), but I'll work on changing the implementation to handle this equality delete case so that the behavior is (a).

wypoon commented on gradle.properties:
```diff
@@ -32,4 +32,4 @@ org.gradle.parallel=true
 # explicitly disable the configuration cache
 org.gradle.configuration-cache=false
-org.gradle.jvmargs=-Xmx1024m
+org.gradle.jvmargs=-Xmx2048m
```
Java CI failed with "Java heap space" error.

manuzhang replied:
Did it just fail here? What's changed?

wypoon replied:
It could be an intermittent issue. After one of my pushes, the CI that ran failed. That is the reason I made this change, to get it to pass. I can try reverting this gradle change and see if the CI passes. If it does, that means it is an intermittent problem.

wypoon replied:
I reverted the gradle change and the CI passed. It appears to be an inconsistent issue.

wypoon commented on data/src/main/java/org/apache/iceberg/data/DeleteFilter.java:
```diff
@@ -197,31 +197,31 @@ record -> deleteSet.contains(projectRow.wrap(asStructLike(record)));
     return eqDeleteRows;
   }
 
+  // Predicate to test whether a row has not been deleted by equality deletes
+  public Predicate<T> eqDeletedRowFilter() {
+    return isEqDeleted().negate();
+  }
```
This is only used in one place, `ColumnarBatchReader.ColumnarBatchLoader::applyEqDelete(ColumnarBatch)`.
I propose deprecating this and changing the call to the negation of `isEqDeleted()` instead. `isEqDeleted()` is more useful. I'll do that in a separate PR.

wypoon commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/source/ChangelogRowReader.java:
```diff
@@ -112,13 +149,62 @@ private CloseableIterable<InternalRow> openChangelogScanTask(ChangelogScanTask t
-    return deletes.filter(rows(task, deletes.requiredSchema()));
+    int[] indexes = indexesInRow(deletes.requiredSchema());
+
+    return CloseableIterable.transform(
+        deletes.filter(rows(task, deletes.requiredSchema())),
+        row -> {
+          InternalRow expectedRow = new GenericInternalRow(columns.length);
+
+          for (int i = 0; i < columns.length; i++) {
+            expectedRow.update(i, row.get(indexes[i], sparkColumnTypes[i]));
+          }
+
+          return expectedRow;
+        });
```
When `DeleteFilter::filter` is called, if there are deletes to be applied, the `requiredSchema` could have additional columns beyond the requested schema (e.g., `_pos` is added if there are positional deletes to be applied). Thus the `InternalRow` could have more columns than the requested schema, and we need to exclude the additional columns and only keep the ones for the requested schema.

dramaticlly replied:
yeah I realized this as well when reading from RowDataReader, looks like if there's position delete present it will always add `pos` metadata column as required in https://github.com/apache/iceberg/blob/684f7a767c2c216a402b60b73d2d55ef605921a0/data/src/main/java/org/apache/iceberg/data/DeleteFilter.java#L260-L263, as well as equality delete id if equality delete are present.

dramaticlly replied:
Speak of which, i saw this update to `InternalRow` has been repeated here and below for deletedDataFileScanTask, do you know if we can put it into a method instead?

wypoon replied:
The code looks the same, but it is not the same transform as it is a closure, and `indexes` in the each closure is different, as `indexes` is a function of the `SparkDeleteFilter` in each case and that is different.
I didn't try too hard to avoid the code repetition here, as I didn't feel it was worth the effort.

wypoon replied:
I'm adding a method `projectRow`, so that the transform can be written as `row -> projectRow(indexes)`.

wypoon commented:
@pvary I have fixed the implementation so that existing deletes are applied before new deletes are emitted. I have fixed the test case accordingly. (I also renamed `testFlinkScenario1` and t`estFlinkScenario2`.) I think things work as you expect now.

pvary commented on core/src/main/java/org/apache/iceberg/BaseIncrementalChangelogScan.java:
```diff
@@ -63,33 +60,43 @@ protected CloseableIterable<ChangelogScanTask> doPlanFiles(
+                    manifestGroup = manifestGroup.ignoreResiduals();
+                  }
+
+                  if (dataManifests.size() > 1 && shouldPlanWithExecutor()) {
```
Do we want to plan with an executor if we have many equality delete files?

wypoon replied:
The parallelism in `ManifestGroup::plan` is at the level of data manifest files, i.e., the unit of work is processing a data manifest file. So asking for planning with an executor service if there are many equality delete files does not help if there is only one data manifest file.

pvary replied:
How does the planning handle when the snapshot only has a few equality delete files?

wypoon replied:
As I wrote, a unit of work is processing a data manifest file. If we have multiple data manifest files (`dataManifests.size() > 1` is true), they can be parallelized. For each unit of work, we produce scan tasks for the manifest entries in the manifest. Each manifest entry (a data file) can have some number of delete files (whether position or equality deletes) associated with it, but that number doesn't affect the parallelizability of the work.
I'm really not sure I understand your question.

pvary replied:
I am concerned about a situation when there is no new data files in a snapshot.
Let's say, we only delete a single row with either a positional or equality delete. How do we handle this situation? Especially when we have multiple of these delete files, but no data files at all?

Thanks for your commitment and efforts. Great to see this feature progressing!

wypoon replied:
I see where you're coming from.
Before this change, there is only one call to `ManifestGroup:plan`. The `ManifestGroup` contains manifests from all the snapshots in the range of interest, only manifest entries added in one of these snapshots are considered and manifests with only existing entries are ignored; then one call to `plan` generates the scan tasks. This approach only works for the copy-on-write case where there are no delete files; in the copy-on-write case, when deletes occur, new data files are always added and so will show up in the `ManifestGroup`. When we have delete files, a situation as you describe can easily arise, where there is no new data file added in a snapshot, only new delete files. That is why we need a different approach.
With this change, there is a `ManifestGroup` created for each snapshot in the range of interest and a call to `ManifestGroup:plan` for it; then the `CloseableIterable` of scan tasks for all of these are concatenated into one `CloseableIterable`. We no longer ignore data manifests with existing entries; in fact, we have to consider them. In a snapshot with no added data files, just added delete files, we consider the existing data files and see if any deletes apply to them.
Now, each call to `ManifestGroup:plan` is still only parallelizable depending on the number of data manifest files in that `ManifestGroup`. There is additional parallelism possible, but I'm not aware that we have the API for it, which is since we are making multiple `ManifestGroup:plan` calls (but on different `ManifestGroup` instances), to distribute each unit of work from every call to one executor service. That is not so easy to do with the API we have, I think.

pvary replied:
Thanks!
I missed that we are iterating through the manifest files, and not the elements of the files.

pvary commented on core/src/main/java/org/apache/iceberg/BaseIncrementalChangelogScan.java:
```diff
@@ -63,33 +60,43 @@ protected CloseableIterable<ChangelogScanTask> doPlanFiles(
+                  long snapshotId = snapshot.snapshotId();
+                  return manifestGroup.plan(
+                      new CreateDataFileChangeTasks(
+                          snapshotId, snapshotOrdinals, deleteFileToSnapshotOrdinal));
```
I'm not sure that we can filter with the current snapshotId. I think we need to filter with the `toSnapshotId` of the scan, so we don't emit records which are added and deleted when the `IncrementalScan` is for multiple snapshots.

I have asked the corresponding question on the dev list thread, but for reference:

> What is the expected behaviour when the `IncrementalScan` is created for not a single snapshot, but for multiple snapshots?
> - S1 added PK1-V1
> - S2 updated PK1-V1 to PK1-V1b (removed PK1-V1 and added PK1-V1b)
> - S3 updated PK1-V1b to PK1-V1c (removed PK1-V1b and added PK1-V1c)
> 
> Let's say we have IncrementalScan.fromSnapshotInclusive(S1).toSnapshot(S3).
> Or we need to return:
> (a)
> - PK1,V1c,INSERTED
> 
> Or is it ok, to return:
> (b)
> - PK1,V1,INSERTED
> - PK1,V1,DELETED
> - PK1,V1b,INSERTED
> - PK1,V1b,DELETED
> - PK1,V1c,INSERTED

I think the (a) is the correct behaviour.

wypoon replied:
Ah, in this case, (b) is the correct behavior. The changelog scan is an incremental scan over multiple snapshots in a range, and should emit the changes for each snapshot. This is the current behavior for the supported case, which is copy-on-write. What you are seeking are the net changes, which is functionality that is also supported by Spark, and built on top of the changelog scan. This uses `ChangelogIterator.removeNetCarryovers`. This functionality is exposed in the Spark procedure, `create_changelog_view`. (Of course, one can also use it programmatically.)

pvary replied:
As agreed on the dev list, let's proceed with the snapshot-by-snapshot reader for now, and decide on the net_changes solution later

pvary commented on core/src/main/java/org/apache/iceberg/BaseIncrementalChangelogScan.java:
```diff
@@ -133,51 +131,149 @@ private static Map<Long, Integer> computeSnapshotOrdinals(Deque<Snapshot> snapsh
+              entry -> {
+                long entrySnapshotId = entry.snapshotId();
+                DataFile dataFile = entry.file().copy(context.shouldKeepStats());
+                DeleteFile[] addedDeleteFiles = filterAdded(context.deletes().forEntry(entry));
```
We can reuse `context.deletes().forEntry(entry)` later, it might worth to add a variable for it and reuse later

wypoon replied:
You are right.

pvary commented on core/src/main/java/org/apache/iceberg/BaseIncrementalChangelogScan.java:
```diff
@@ -133,51 +131,149 @@ private static Map<Long, Integer> computeSnapshotOrdinals(Deque<Snapshot> snapsh
+                        "Unexpected entry status: " + entry.status());
+                }
+              });
+      return CloseableIterable.filter(tasks, task -> !(task instanceof DummyChangelogScanTask));
```
Would `task != DummyChangelogScanTask.INSTANCE` work?

wypoon replied:
Yes, that works too. `instanceof` was just how I thought of it logically, and as I understand it, `instanceof` is highly optimized so performance is probably no worse.

pvary replied:
In my experience using `instanceof` is a sign of bad code when we don't use inheritance/object oriented principles correctly. I understand it that this is not the case here, but when reading the code, every new reader must learn/understand it, which is a small extra effort from every future developer. Also I would be surprised if the performance would be better than default implementation of the `equals` 😄

pvary replied:
Thanks for changing!

pvary commented:
Could we add some tests for the `BaseIncrementalChangelogScan` directly?
It would be good if we don't depend on Spark to test core functionality.
Thanks, Peter

dramaticlly commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/source/ChangelogRowReader.java:
```diff
@@ -112,13 +149,62 @@ private CloseableIterable<InternalRow> openChangelogScanTask(ChangelogScanTask t
+    SparkDeleteFilter existingDeletes =
+        new SparkDeleteFilter(filePath, task.existingDeletes(), counter());
+    SparkDeleteFilter newDeletes = new SparkDeleteFilter(filePath, task.addedDeletes(), counter());
+    Schema schema1 = existingDeletes.requiredSchema();
+    Schema schema2 = newDeletes.requiredSchema();
+    Schema requiredSchema = TypeUtil.join(schema1, schema2);
```
can we inline this to 
```
 Schema requiredSchema = TypeUtil.join(existingDeletes.requiredSchema(), newDeletes.requiredSchema());
```

wypoon replied:
Sure, that is a minor thing.

dramaticlly commented on core/src/main/java/org/apache/iceberg/BaseIncrementalChangelogScan.java:
```diff
@@ -133,51 +131,149 @@ private static Map<Long, Integer> computeSnapshotOrdinals(Deque<Snapshot> snapsh
+
+    private DeleteFile[] filterAdded(DeleteFile[] deleteFiles) {
+      return FluentIterable.from(deleteFiles)
+          .filter(
+              deleteFile ->
+                  deleteFileToSnapshotOrdinal.get(deleteFile.path().toString()) == changeOrdinal)
+          .toArray(DeleteFile.class);
+    }
 
-    CreateDataFileChangeTasks(Deque<Snapshot> snapshots) {
-      this.snapshotOrdinals = computeSnapshotOrdinals(snapshots);
+    private DeleteFile[] filterExisting(DeleteFile[] deleteFiles) {
+      return FluentIterable.from(deleteFiles)
+          .filter(
+              deleteFile ->
+                  deleteFileToSnapshotOrdinal.get(deleteFile.path().toString()) < changeOrdinal)
+          .toArray(DeleteFile.class);
     }
```
Can we use a comparator here to highlight the difference between added deletes and existing deletes? I think you can compare it with the snapshot ordinal but it's a bit hard to see as the only difference is equality and less than.

dramaticlly commented on core/src/main/java/org/apache/iceberg/BaseIncrementalChangelogScan.java:
```diff
@@ -133,51 +131,149 @@ private static Map<Long, Integer> computeSnapshotOrdinals(Deque<Snapshot> snapsh
+      return FluentIterable.from(deleteFiles)
+          .filter(
+              deleteFile ->
+                  deleteFileToSnapshotOrdinal.get(deleteFile.path().toString()) < changeOrdinal)
```
Can we use a comparator here and line 190 to highlight the difference between added deletes and existing deletes? I think you can compare it with the snapshot ordinal but it's a bit hard to see as the only difference is equality and less than.

wypoon replied:
I'm not sure I understand the suggestion. Perhaps you can elaborate?
I thought the code here is quite clear and the intention is apparent from the method name. Thus `filterAdded` filters for delete files whose snapshot ordinal is the same as (`==`) `changeOrdinal`, while `filterExisting` filters for delete files whose snapshot ordinal comes before (`<`) `changeOrdinal`.

wypoon replied:
These methods have been removed in the latest version, so this is no longer relevant.

dramaticlly reviewed (commented):
thank you @wypoon for your change. Originally I was planning to help on this by starting with #10924  but I do not have enough expertise on how to update the DeleteFilter and BaseReader. I am glad that now we can focus on your PR to support CDC scan for MOR tables!

wypoon commented:
@pvary and @dramaticlly, thank you both for reviewing. I am busy with some other work at the moment, but I'll return to this by next week.

wypoon commented on core/src/main/java/org/apache/iceberg/BaseIncrementalChangelogScan.java:
```diff
@@ -63,33 +60,43 @@ protected CloseableIterable<ChangelogScanTask> doPlanFiles(
-    if (newDataManifests.size() > 1 && shouldPlanWithExecutor()) {
-      manifestGroup = manifestGroup.planWith(planExecutor());
-    }
+    // map of delete file to the snapshot where the delete file is added
+    // the delete file is keyed by its path, and the snapshot is represented by the snapshot ordinal
+    Map<String, Integer> deleteFileToSnapshotOrdinal =
+        computeDeleteFileToSnapshotOrdinal(changelogSnapshots, snapshotOrdinals);
```
I suspect that when the dust is settled on the V3 spec, there will be changes needed to the changelog implementation, and not just this part. I agree that we should try to keep the proposed changes in mind, but I don't think we should hold up implementing something for V2. We have to deal with V3 changes when that comes.
Right now, the scan tasks we have to work with take `DeleteFile[]` parameters, as they are designed for V2. So for V2 this map makes sense for use in filtering the `DeleteFile`s to pass to those scan task classes.

dramaticlly replied:
yeah I agree the assessment

wypoon commented:
I did not have time to come back to this work this week as I hoped, but I think I should have some time next week.

wypoon commented on spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestChangelogReader.java:
```diff
@@ -191,39 +214,359 @@ public void testMixDeleteAndInsert() throws IOException {
 
-    // order by the change ordinal
+    // order by change ordinal, change type, data, id
     rows.sort(
         (r1, r2) -> {
           if (r1.getInt(3) != r2.getInt(3)) {
             return r1.getInt(3) - r2.getInt(3);
+          } else if (!r1.getUTF8String(2).equals(r2.getUTF8String(2))) {
+            return r1.getUTF8String(2).compareTo(r2.getUTF8String(2));
+          } else if (!r1.getUTF8String(1).equals(r2.getUTF8String(1))) {
+            return r1.getUTF8String(1).compareTo(r2.getUTF8String(1));
           } else {
             return r1.getInt(0) - r2.getInt(0);
           }
         });
```
Thanks for the suggestion!

wypoon commented:
@pvary I added some tests in `TestBaseIncrementalChangelogScan`. However, at that level, we can only check what scan tasks (`AddedRowsScanTask`, `DeletedRowsScanTask`, `DeletedDataFileScanTask`) are generated. We cannot actually check the kind of results as we can in the Spark `TestChangelogReader`. Right now, the only consumer of `IncrementalChangelogScan` is Spark, so some of the functionality needs to be tested through Spark.

pvary commented:
> @pvary I added some tests in `TestBaseIncrementalChangelogScan`. However, at that level, we can only check what scan tasks (`AddedRowsScanTask`, `DeletedRowsScanTask`, `DeletedDataFileScanTask`) are generated. We cannot actually check the kind of results as we can in the Spark `TestChangelogReader`. Right now, the only consumer of `IncrementalChangelogScan` is Spark, so some of the functionality needs to be tested through Spark.

Makes sense. Give me some time to review the added tests.

pvary commented on core/src/test/java/org/apache/iceberg/TestBaseIncrementalChangelogScan.java:
```diff
@@ -132,6 +131,139 @@ public void testFileDeletes() {
   }
 
+  @TestTemplate
+  public void testPositionDeletes() {
```
nice stuff!
Could we have 2 more test cases?
1. It would be nice to test the existing deletes scenario
2. It would be nice to test when the scan returns result for multiple snapshots

Thanks,
Peter

wypoon replied:
Thank you for the review.
I consolidated `testPositionDeletes`, `testEqualityDeletes` and `testMixOfPositionAndEqualityDeletes` into one test to reduce repetition as well as to test a scan with multiple snapshots. I also added tests for scan tasks with existing deletes.
In the process, I discovered a bug in some edge cases and need to work on a fix before I push an update.

pvary replied:
@wypoon: We usually try to create different tests for different test cases. So your original approach seems nice. Could we remove the code duplication with parameterized tests, or externalizing the duplicated code to test methods?

wypoon replied:
Please look at the new consolidated test, `testRowDeletes`. I think that at the level at which we can test `BaseIncrementalChangelogScan`, it doesn't matter whether delete files are position delete files or equality delete files (they only matter when we actually apply the deletes, which doesn't happen at this level), so `testPositionDeletes` and `testEqualityDeletes` don't really test different cases.

pvary replied:
Thanks!
I like your new tests!

wypoon commented on core/src/main/java/org/apache/iceberg/BaseIncrementalChangelogScan.java:
```diff
@@ -133,51 +128,124 @@ private static Map<Long, Integer> computeSnapshotOrdinals(Deque<Snapshot> snapsh
+                long entrySnapshotId = entry.snapshotId();
+                DataFile dataFile = entry.file().copy(context.shouldKeepStats());
+                DeleteFile[] deleteFiles = context.deletes().forEntry(entry);
+                List<DeleteFile> added = Lists.newArrayList();
+                List<DeleteFile> existing = Lists.newArrayList();
+                for (DeleteFile deleteFile : deleteFiles) {
+                  if (sequenceNumber == deleteFile.dataSequenceNumber()) {
+                    added.add(deleteFile);
+                  } else {
+                    existing.add(deleteFile);
+                  }
+                }
+                DeleteFile[] addedDeleteFiles = added.toArray(new DeleteFile[0]);
+                DeleteFile[] existingDeleteFiles = existing.toArray(new DeleteFile[0]);
```
Previously, I used a map from delete files to the snapshot where it is added, to determine the `addedDeleteFiles` and `existingDeleteFiles`. However, this map is computed for the snapshots in the range of interest (for the `BaseIncrementalChangelogScan`), and so there may not be an entry in the map for a delete file if it is added in a snapshot before this range. This causes issues in the `FluentIterable::filter(Predicate)` I was using to filter the delete files. This approach is simpler, and turns out to be what @manuzhang used his https://github.com/apache/iceberg/pull/9888. Hat tip to Manu.

wypoon replied:
@pvary this was the bug I mentioned that I needed to fix before pushing a change with additional tests in `TestBaseIncrementalChangelogScan`.
In `TestBaseIncrementalChangelogScan`, in `testDeletingDataFileWithExistingDeletes` and `testDeletingRowsInDataFileWithExistingDeletes`, if the `IncrementalChangelogScan` was from `snap1` to `snap3`, the bug was obscured; but with a scan from `snap2` to `snap3`, those tests revealed the bug.

dramaticlly commented on core/src/test/java/org/apache/iceberg/TestBaseIncrementalChangelogScan.java:
```diff
@@ -132,6 +131,175 @@ public void testFileDeletes() {
+    assertThat(t1.file().path()).as("Data file must match").isEqualTo(FILE_B.path());
+    assertThat(t1.addedDeletes().get(0).path())
+        .as("Added delete file must match")
+        .isEqualTo(FILE_B_DELETES.path());
+    assertThat(t1.existingDeletes()).as("Must be no existing deletes").isEmpty();
+
+    DeletedRowsScanTask t2 = (DeletedRowsScanTask) tasks.get(1);
+    assertThat(t2.changeOrdinal()).as("Ordinal must match").isEqualTo(1);
+    assertThat(t2.commitSnapshotId()).as("Snapshot must match").isEqualTo(snap3.snapshotId());
+    assertThat(t2.file().path()).as("Data file must match").isEqualTo(FILE_C.path());
+    assertThat(t2.addedDeletes().get(0).path())
+        .as("Added delete file must match")
+        .isEqualTo(FILE_C2_DELETES.path());
+    assertThat(t2.existingDeletes()).as("Must be no existing deletes").isEmpty();
+
+    DeletedRowsScanTask t3 = (DeletedRowsScanTask) tasks.get(2);
+    assertThat(t3.changeOrdinal()).as("Ordinal must match").isEqualTo(2);
+    assertThat(t3.commitSnapshotId()).as("Snapshot must match").isEqualTo(snap4.snapshotId());
+    assertThat(t3.file().path()).as("Data file must match").isEqualTo(FILE_A2.path());
+    assertThat(t3.addedDeletes().size()).as("Number of added delete files must match").isEqualTo(2);
+    assertThat(t3.addedDeletes().get(0).path())
+        .as("Added delete file must match")
+        .isEqualTo(FILE_A2_DELETES.path());
+    assertThat(t3.addedDeletes().get(1).path())
+        .as("Added delete file must match")
+        .isEqualTo(FILE_A_DELETES.path());
+    assertThat(t3.existingDeletes()).as("Must be no existing deletes").isEmpty();
+
+    DeletedRowsScanTask t4 = (DeletedRowsScanTask) tasks.get(3);
+    assertThat(t4.changeOrdinal()).as("Ordinal must match").isEqualTo(2);
+    assertThat(t4.commitSnapshotId()).as("Snapshot must match").isEqualTo(snap4.snapshotId());
+    assertThat(t4.file().path()).as("Data file must match").isEqualTo(FILE_A.path());
+    assertThat(t4.addedDeletes().size()).as("Number of added delete files must match").isEqualTo(2);
+    assertThat(t4.addedDeletes().get(0).path())
+        .as("Added delete file must match")
+        .isEqualTo(FILE_A2_DELETES.path());
+    assertThat(t4.addedDeletes().get(1).path())
+        .as("Added delete file must match")
+        .isEqualTo(FILE_A_DELETES.path());
+    assertThat(t4.existingDeletes()).as("Must be no existing deletes").isEmpty();
```
I believe this is the core of the test I am looking for! I think there's some feature we can leverage in assertJ so I took the stub to organize this in a little different way.

I do notice that the we use the file path instead of object itself for asserting equality on data and delete files,  probably because once added to the table, its data and file sequence number are inherited from manifest entry. I think we can probably reuse this `FILE_COMPARISON_CONFIG` from another class [`TestManifestReader`](https://github.com/apache/iceberg/blob/5ea78e3fbe5d8c9c846a1b86bcd2b77d13a31acd/core/src/test/java/org/apache/iceberg/TestManifestReader.java#L40)

```java
  private static final RecursiveComparisonConfiguration FILE_COMPARISON_CONFIG =
      RecursiveComparisonConfiguration.builder()
          .withIgnoredFields(
              "dataSequenceNumber", "fileOrdinal", "fileSequenceNumber", "fromProjectionPos")
          .build();
```

complete code 
- group assertion around changelogScanTask
- ensure all are actually instance of DeletedRowsScanTask and assert on its data and existing/added delete files
- leverage FILE_COMPARISON_CONFIG for asserting equality
```java
    assertThat(tasks)
        .extracting(ChangelogScanTask::changeOrdinal)
        .as("Ordinal must match")
        .containsExactly(0, 1, 2, 2);

    assertThat(tasks)
        .extracting(ChangelogScanTask::commitSnapshotId)
        .as("Snapshot must match")
        .containsExactly(
            snap2.snapshotId(), snap3.snapshotId(), snap4.snapshotId(), snap4.snapshotId());

    assertThat(tasks)
        .extracting(ChangelogScanTask::operation)
        .as("Operation must match")
        .containsExactly(
            ChangelogOperation.DELETE,
            ChangelogOperation.DELETE,
            ChangelogOperation.DELETE,
            ChangelogOperation.DELETE);

    List<DeletedRowsScanTask> deletedRowsScanTasks =
        tasks.stream()
            .filter(t -> t instanceof DeletedRowsScanTask)
            .map(t -> (DeletedRowsScanTask) t)
            .collect(Collectors.toList());
    assertThat(deletedRowsScanTasks).as("Must have 4 DeletedRowsScanTask").hasSize(4);
    assertThat(deletedRowsScanTasks)
        .extracting(DeletedRowsScanTask::existingDeletes)
        .as("Existing deletes shall be empty")
        .allMatch(List::isEmpty);

    DeletedRowsScanTask t1 = deletedRowsScanTasks.get(0);
    assertThat(t1.file())
        .usingRecursiveComparison(FILE_COMPARISON_CONFIG)
        .as("Data file must match")
        .isEqualTo(FILE_B);
    assertThat(t1.addedDeletes().get(0))
        .usingRecursiveComparison(FILE_COMPARISON_CONFIG)
        .as("Added delete file must match")
        .isEqualTo(FILE_B_DELETES);

    DeletedRowsScanTask t2 = deletedRowsScanTasks.get(1);
    assertThat(t2.file())
        .usingRecursiveComparison(FILE_COMPARISON_CONFIG)
        .as("Data file must match")
        .isEqualTo(FILE_C);
    assertThat(t2.addedDeletes().get(0))
        .usingRecursiveComparison(FILE_COMPARISON_CONFIG)
        .as("Added delete file must match")
        .isEqualTo(FILE_C2_DELETES);

    DeletedRowsScanTask t3 = deletedRowsScanTasks.get(2);
    assertThat(t3.file())
        .usingRecursiveComparison(FILE_COMPARISON_CONFIG)
        .as("Data file must match")
        .isEqualTo(FILE_A2);
    assertThat(t3.addedDeletes()).as("Number of added delete files must match").hasSize(2);
    assertThat(t3.addedDeletes().get(0))
        .usingRecursiveComparison(FILE_COMPARISON_CONFIG)
        .as("Added delete file must match")
        .isEqualTo(FILE_A2_DELETES);
    assertThat(t3.addedDeletes().get(1))
        .usingRecursiveComparison(FILE_COMPARISON_CONFIG)
        .as("Added delete file must match")
        .isEqualTo(FILE_A_DELETES);

    DeletedRowsScanTask t4 = deletedRowsScanTasks.get(3);
    assertThat(t4.file())
        .usingRecursiveComparison(FILE_COMPARISON_CONFIG)
        .as("Data file must match")
        .isEqualTo(FILE_A);
    assertThat(t4.addedDeletes()).as("Number of added delete files must match").hasSize(2);
    assertThat(t4.addedDeletes().get(0))
        .usingRecursiveComparison(FILE_COMPARISON_CONFIG)
        .as("Added delete file must match")
        .isEqualTo(FILE_A2_DELETES);
    assertThat(t4.addedDeletes().get(1))
        .usingRecursiveComparison(FILE_COMPARISON_CONFIG)
        .as("Added delete file must match")
        .isEqualTo(FILE_A_DELETES);
```

wypoon replied:
Thanks for thinking about this and the idea. I learned something about AssertJ from you!
As I see it, however, the example you show does not actually result in reduction in code, just that it is organized differently. Also, as far as using the `RecursiveComparisonConfiguration` goes, that part really does not reduce code at all. Conceptually, we want to check what the deletes/addedDeletes/existingDeletes are, and it is enough to check that the file path matches with the path of the `DeleteFile`s we expect. In other words,
```
    assertThat(t4.addedDeletes().get(0).path())
        .isEqualTo(FILE_A2_DELETES.path());
```
is just as good as
```
    assertThat(t4.addedDeletes().get(0))
        .usingRecursiveComparison(FILE_COMPARISON_CONFIG)
        .isEqualTo(FILE_A2_DELETES);
```
and is slightly shorter.

dramaticlly replied:
In that sense I believe we can probably simplify assertion on addedDeletes to be this (with code reduction)

```java
assertThat(t3.addedDeletes())
    .hasSize(2)
    .extracting(DeleteFile::path)
    .as("Added delete file must match")
    .containsExactly(
            FILE_A2_DELETES.path(),
            FILE_A_DELETES.path());
```

wypoon replied:
That sounds good. Thanks!

pvary commented on core/src/test/java/org/apache/iceberg/TestBaseIncrementalChangelogScan.java:
```diff
@@ -132,6 +131,175 @@ public void testFileDeletes() {
+    assertThat(t3.commitSnapshotId()).as("Snapshot must match").isEqualTo(snap4.snapshotId());
+    assertThat(t3.file().path()).as("Data file must match").isEqualTo(FILE_A2.path());
+    assertThat(t3.addedDeletes().size()).as("Number of added delete files must match").isEqualTo(2);
+    assertThat(t3.addedDeletes().get(0).path())
+        .as("Added delete file must match")
+        .isEqualTo(FILE_A2_DELETES.path());
+    assertThat(t3.addedDeletes().get(1).path())
+        .as("Added delete file must match")
+        .isEqualTo(FILE_A_DELETES.path());
```
would this be flaky? Is there a guarantee for the order of the files?

dramaticlly replied:
if I understand this correctly, the test result seem to be consistent here as when build DeleteFileIndex, we always concat equality delete (A2_DELETES) before position delete (A_DELETES) for given data file entry in https://github.com/apache/iceberg/blob/9dbfbbb203c007592b8ba5f4816925043c08923a/core/src/main/java/org/apache/iceberg/DeleteFileIndex.java#L139 and array ordering seem to be maintained through out the code. 


IMO, this is not necessarily something we need to assert the ordering on

wypoon replied:
The test is not flaky, but you have a good point. We need not depend on the order. We can follow a variant of @dramaticlly's suggestion below:
```
    assertThat(t3.addedDeletes())
        .hasSize(2)
        .extracting(DeleteFile::path)
        .as("Added delete file must match")
        .containsExactlyInAnyOrder(
                FILE_A2_DELETES.path(),
                FILE_A_DELETES.path());
```

pvary commented on core/src/test/java/org/apache/iceberg/TestBaseIncrementalChangelogScan.java:
```diff
@@ -132,6 +131,175 @@ public void testFileDeletes() {
+    Snapshot snap3 = table.currentSnapshot();
+
+    // mix of position and equality deletes
+    table.newRowDelta().addDeletes(FILE_A_DELETES).addDeletes(FILE_A2_DELETES).commit();
```
One more interesting question is:
- How we will handle the compaction commits? Is there a better way than emit every insert/delete record? Could we skip it?

Is it trivial to fix in this PR, or is that need another PR?

wypoon replied:
I'm not sure I understand your question. Are you referring to outputting *net changes*? If so, I think that should be a separate PR. I think for that, we should introduce an option for it.
If you're referring to something else, can you please explain?

wypoon replied:
ps. Net changes is something a user might want regardless of whether merge-on-read or copy-on-write is used, so is an orthogonal concern.

pvary replied:
This is the question about how you see the compaction where we rewrite the files with the expressed goal to keep the actual content unchanged.
- Is this a change where we remove and add files which are coincidentally contain the same data?
- Is this a change where no data has changed?

I see the reasoning behind both cases, but using the first approach would most probably make the changelog scan unfesable for cases where compaction happens on the table.

wypoon replied:
Ah, you mean compaction as in rewriting data files.
In this case, compaction does not result in any inserts or deletes. (I believe this is what we both expect.) Please see `testDataRewritesAreIgnored` that I added in `TestChangelogTable` in this PR. This test is in spark-extensions because I want to call the rewrite_data_files Spark procedure.

pvary replied:
Cool!
Thanks for the explanation!

pvary commented:
This looks good to me.
@dramaticlly: Any more comments, before we try to involve the guys who are more experienced with the core/spark parts?

pvary commented:
@RussellSpitzer, @aokolnychyi: would it be possible to take a look at this PR? I did my best to review it, but I'm not an expert in this part of the code.
Thanks, Peter

wypoon commented:
@flyrain you worked on this area and implemented some of the changelog support; can you please review?

aokolnychyi commented:
Will start looking into this PR today and should be able to finish over the weekend.

wypoon commented:
@flyrain and @aokolnychyi as you have both expressed interest to review this, I hope you can find time to do so soon! Thank you!

aokolnychyi commented on core/src/main/java/org/apache/iceberg/BaseIncrementalChangelogScan.java:
```diff
@@ -63,33 +61,39 @@ protected CloseableIterable<ChangelogScanTask> doPlanFiles(
+                  List<ManifestFile> deleteManifests = snapshot.deleteManifests(table().io());
+
+                  ManifestGroup manifestGroup =
+                      new ManifestGroup(table().io(), dataManifests, deleteManifests)
```
This will cause a substantial performance hit as we will scan all data and delete manifests that match the filter for each changelog snapshot, instead of opening only newly added manifests before. We may have to do that for deletes anyway, but I wonder about data manifests. Let me think a bit.

wypoon replied:
I went snapshot by snapshot as it was the easiest to reason with. For sure we need to consider more than just newly added manifests. For each snapshot, we need to consider existing data files if there are new deletes applicable to them.

aokolnychyi commented:
I went through some of my old notes, which we should discuss.

We have the following tasks right now:
- `AddedRowsScanTask` (added data file + deletes that happened within the same snapshot).
- `DeletedDataFileScanTask` (removed data file + deletes that applied to it before it was removed).
- `DeletedRowsScanTask` (data file that was affected by a delete file (if we resolve equality deletes or we had position deletes) + historic deletes that were there before + new deletes added in this snapshot).

First of all, we need to discuss the expected behavior:
- Do we want to resolve equality deletes and map them into data files? Or should we add a new task and output the content of equality delete files? I'd say we should support both options (without resolving by default?).
- What if a snapshot adds a new delete file for the already deleted record? That can happen both with position and equality deletes. For instance, if two position-based delete operations remove the same set of records, both of them will succeed. Producing a precise CDC log would require reading all historic deletes, which may be unnecessary expensive in some cases. I'd say this should be configurable as well.

If we want to make resolving equality deletes optional, we need `AddedEqualityDeletesScanTask`. We discussed this for Flink CDC use cases. It is going to be costly at planning and query time to apply equality deletes to data files to get removed records. As long as the equality delete persists the entire row or the caller is OK with only equality columns, it should be fine to output the content of equality delete files as is.

I agree with the idea of iterating snapshot by snapshot. I wonder if we can optimize it, though.

**AddedRowsScanTask**

We don’t have to look up historic deletes as the file was just added, except the position deletes added in the same snapshot. For each added data file, it is enough to look up matching position deletes in `DeleteFileIndex` built from delete manifests added in this snapshot.

**DeletedDataFileScanTask**

We have to build `DeleteFileIndex` that includes all historic deletes for removed data files. We can optimize this step by reading delete manifests only for the affected partitions. We can create a `PartitionMap` predicate from new data files and use it while reading delete manifests. We can supplement that with a predicate on `referencedDataFile` in the future. Historic delete files that are not affecting deleted data file partitions can be discarded.

**DeletedRowsScanTask**

For each position delete file and each equality delete file (if those must be resolved), find matching data files and historic deletes (if configured to resolve double deletes). We can still build a partition predicate from all delete files that were added in this snapshot. We can use that predicate to prune the set of data manifests for that snapshot. For each newly added delete file, output a task for each affected data file (one delete file can match multiple data files).

**AddedEqualityDeletesScanTask**

Output each added equality delete file as a separate task without matching them to data files (if configured).

aokolnychyi commented:
What I wrote above is more complicated, we need to understand if that complexity will be justified. I am not sure yet.

pvary commented:
> First of all, we need to discuss the expected behavior:
> 
> * Do we want to resolve equality deletes and map them into data files? Or should we add a new task and output the content of equality delete files? I'd say we should support both options (without resolving by default?).

If the table is written by Flink, then for Flink CDC streaming read the lazy solution (not resolving the equality deletes) would be enough. If there is another writer for the table which creates non Flink conform equality delete files, or the user wants a retracting CDC stream when the table was written by an upsert CDC stream, then the equality delete resolution is still needed.

> * What if a snapshot adds a new delete file for the already deleted record? That can happen both with position and equality deletes. For instance, if two position-based delete operations remove the same set of records, both of them will succeed. Producing a precise CDC log would require reading all historic deletes, which may be unnecessary expensive in some cases. I'd say this should be configurable as well.

I think that applying delete files is less costly, so I would stick to the theoretically correct solution, and apply previously added delete files to produce the precise CDC log

wypoon commented:
> For instance, if two position-based delete operations remove the same set of records, both of them will succeed.

I was not aware of this. Is this specifically the case of two concurrent delete operations starting from the same base snapshot, both removing the same set of records, and one winning the commit race, but the other would still succeed in committing?
Or is this more general, where a delete that deletes already deleted records is allowed to do so?

Let's say we are in the former case, then I'd say that for the purpose of CDC, the second commit would contain no changes, so the changelog should not write anything for it.

aokolnychyi commented:
Coming back to some questions above.

> Do we want to resolve equality deletes and map them into data files? Or should we add a new task and output the content of equality delete files?

While it would be useful to stream out ingested CDC log from Flink without applying equality deletes, we shouldn't probably target that for now. If we start outputting equality deletes directly, our changelog may not be accurate. What if we upsert a record that didn't exist? What if an equality delete applied to 10 data records? We can find reasonable behavior but let's skip this for now. Let's always resolve and assign equality deletes so that the changelog is precise.

> What if a snapshot adds a new delete file for the already deleted record? That can happen both with position and equality deletes.

There is one more data point here. We are about to introduce sync maintenance for position deletes. This means new delete files will include old + new deleted positions. Therefore, we must always resolve historic deletes.

To sum up, I propose sticking to the existing changelog tasks and always resolving historic deletes to produce a correct changelog. I am concerned about full table scans for each incremental snapshot. I'll look into ideas mentioned above to see if we can optimize that.

wypoon commented:
@aokolnychyi I have rebased the PR on main. How can we move forward with this?
If we agree that the behavior is correct, how about we at least have something that works and optimize it later?

wypoon commented:
@flyrain you had indicated that you were interested in this. Can you please review?

wypoon commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/source/ChangelogRowReader.java:
```diff
@@ -170,4 +240,10 @@ private static Stream<ContentFile<?>> addedRowsScanTaskFiles(AddedRowsScanTask t
+
+  private static Stream<ContentFile<?>> deletedRowsScanTaskFiles(DeletedRowsScanTask task) {
+    DataFile file = task.file();
+    List<DeleteFile> deletes = task.addedDeletes();
```
Hi @guykhazma. I believe you're right. The DeletedRowsScanTask references the existing deletes as well as the added deletes.
Can you elaborate on the situation in which you encounter NPEs?

guykhazma replied:
Hi @wypoon,
We noticed that in a case where we had a custom writer which was creating an overwrite snapshot that contained new delete files as well as adding new data. The delete files in this case were both from the snapshot and existing ones.
